In [ ]:
Individual Assignment 2
Part A - Question 1

In [1]:
import os
import numpy as np
import pandas as pd
from zipfile import ZipFile
from itertools import groupby, combinations
from collections import Counter
from IPython.display import display

In [2]:
input_path = 'C:/Users/shala/Downloads/'

def extract_zip(file_name: str) -> None:
    with ZipFile(file_name, 'r') as zip_file:
        zip_file.extractall()
        print(f'Extracted: {file_name}')

# List of files to extract
files_to_extract = [
    'aisles.csv.zip',
    'departments.csv.zip',
    'order_products__prior.csv.zip',
    'order_products__train.csv.zip',
    'orders.csv.zip',
    'products.csv.zip',
    'sample_submission.csv.zip'
]


In [3]:

# Extract each file
for file in files_to_extract:
    extract_zip(input_path + file)

# Read and display the first few rows of each extracted CSV file
for file in os.listdir():
    if file.endswith('.csv'):
        df = pd.read_csv(file)
        print(f'\n{file}:')
        display(df.head())


Extracted: C:/Users/shala/Downloads/aisles.csv.zip
Extracted: C:/Users/shala/Downloads/departments.csv.zip
Extracted: C:/Users/shala/Downloads/order_products__prior.csv.zip
Extracted: C:/Users/shala/Downloads/order_products__train.csv.zip
Extracted: C:/Users/shala/Downloads/orders.csv.zip
Extracted: C:/Users/shala/Downloads/products.csv.zip
Extracted: C:/Users/shala/Downloads/sample_submission.csv.zip

2015-summary.csv:


DEST_COUNTRY_NAME ORIGIN_COUNTRY_NAME  count
0     United States             Romania     15
1     United States             Croatia      1
2     United States             Ireland    344
3             Egypt       United States     15
4     United States               India     62


aisles new.csv:


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation


aisles.csv:


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation


Assignment 2 - Sponsorships Test Data.csv:


Customer ID Segment Tenure  Invited PBG  Invited W    AUM 2022    AUM 2023
0            1      EA      M            0          0    $356,277    $308,594
1            2      EA      M            1          1  $1,505,866  $1,581,660
2            3     HNW      L            0          0  $9,091,923  $7,597,882
3            4      MA      L            0          0  $1,986,585  $1,200,097
4            5      MA      M            0          1    $221,583    $262,352


Assignment 4 - Guac (1).csv:


Region  Year Quarter   Sales  Price  Harvest  Unnamed: 6  Unnamed: 7  \
0      A  2020      Q1   70.99   8.46     3.22         NaN         NaN   
1      A  2020      Q2  105.56   7.21     3.34         NaN         NaN   
2      A  2020      Q3  101.80   7.64     3.64         NaN         NaN   
3      A  2020      Q4   74.15  10.38     1.52         NaN         NaN   
4      A  2021      Q1   92.36   8.28     3.95         NaN         NaN   

   Unnamed: 8  Unnamed: 9  Unnamed: 10  
0         NaN         NaN          NaN  
1         NaN         NaN          NaN  
2         NaN         NaN          NaN  
3         NaN         NaN          NaN  
4         NaN         NaN          NaN


Assignment 4 - Guac.csv:


Region  Year Quarter   Sales  Price  Harvest  Unnamed: 6  Unnamed: 7  \
0      A  2020      Q1   70.99   8.46     3.22         NaN         NaN   
1      A  2020      Q2  105.56   7.21     3.34         NaN         NaN   
2      A  2020      Q3  101.80   7.64     3.64         NaN         NaN   
3      A  2020      Q4   74.15  10.38     1.52         NaN         NaN   
4      A  2021      Q1   92.36   8.28     3.95         NaN         NaN   

   Unnamed: 8  Unnamed: 9  Unnamed: 10  
0         NaN         NaN          NaN  
1         NaN         NaN          NaN  
2         NaN         NaN          NaN  
3         NaN         NaN          NaN  
4         NaN         NaN          NaN


Assignment 4 - Store-Level Observational Data 2 (1).csv:


Store_number  YEAR  WEEK     F  D  PR  EST_ACV Market_Name  Open  Clsd  \
0        205512    11  1635  NONE  0   0  123.676      BOSTON  1203  9998   
1        205512    11  1636     B  0   1  123.676      BOSTON  1203  9998   
2        205512    11  1637  NONE  0   0  123.676      BOSTON  1203  9998   
3        205512    11  1638  NONE  0   0  123.676      BOSTON  1203  9998   
4        205512    11  1639  NONE  0   0  123.676      BOSTON  1203  9998   

  MskdName  UNITS  DOLLARS  
0  Chain19     29   124.41  
1  Chain19     64   255.36  
2  Chain19     40   171.60  
3  Chain19     38   163.02  
4  Chain19     33   141.57


Assignment 4 - Store-Level Observational Data 2.csv:


Store_number  YEAR  WEEK     F  D  PR  EST_ACV Market_Name  Open  Clsd  \
0        205512    11  1635  NONE  0   0  123.676      BOSTON  1203  9998   
1        205512    11  1636     B  0   1  123.676      BOSTON  1203  9998   
2        205512    11  1637  NONE  0   0  123.676      BOSTON  1203  9998   
3        205512    11  1638  NONE  0   0  123.676      BOSTON  1203  9998   
4        205512    11  1639  NONE  0   0  123.676      BOSTON  1203  9998   

  MskdName  UNITS  DOLLARS  
0  Chain19     29   124.41  
1  Chain19     64   255.36  
2  Chain19     40   171.60  
3  Chain19     38   163.02  
4  Chain19     33   141.57


authors.csv:


author_name  bookID
0  J.K. Rowling/Mary GrandPré       1
1  J.K. Rowling/Mary GrandPré       2
2                 J.K. Rowling       4
3  J.K. Rowling/Mary GrandPré       5
4  J.K. Rowling/Mary GrandPré       8


bank.csv:


age         job  marital  education default  balance housing loan  contact  \
0   59      admin.  married  secondary      no     2343     yes   no  unknown   
1   56      admin.  married  secondary      no       45      no   no  unknown   
2   41  technician  married  secondary      no     1270     yes   no  unknown   
3   55    services  married  secondary      no     2476     yes   no  unknown   
4   54      admin.  married   tertiary      no      184      no   no  unknown   

   day month  duration  campaign  pdays  previous poutcome deposit  
0    5   may      1042         1     -1         0  unknown     yes  
1    5   may      1467         1     -1         0  unknown     yes  
2    5   may      1389         1     -1         0  unknown     yes  
3    5   may       579         1     -1         0  unknown     yes  
4    5   may       673         2     -1         0  unknown     yes


breast_cancer.csv:


id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]


business q1.csv:


publisher language_code  number_of_books  average_rating
0              Penguin Books           eng               11        3.837273
1  Farrar  Straus and Giroux           eng                7        4.211429
2  Houghton Mifflin Harcourt           eng                5        4.496000
3             Broadway Books           eng                5        3.898000
4                Grove Press           eng                5        3.846000


business q2.csv:


author_name first_book_date latest_book_date  first_book_rating  \
0      Bill Bryson      12-03-2002        9/28/1991               4.21   
1  Bryce Courtenay       9/13/2005        9/29/1996               4.35   
2      Don DeLillo      01-01-1994        7/17/1989               3.92   
3    Douglas Adams      08-03-2004        4/30/2002               4.38   
4  Edward P. Jones       6/15/2004        8/29/2006               3.83   

   latest_book_rating  
0                3.93  
1                4.35  
2                3.64  
3                3.93  
4                3.83


business q3.csv:


language_code                                              title  \
0         en-US  The Long Dark Tea-Time of the Soul (Dirk Gentl...   
1           eng  Harry Potter and the Prisoner of Azkaban (Harr...   
2           fre                     La Place de la Concorde Suisse   
3           spa                 El perfume: Historia de un asesino   

   text_reviews_count  
0                1401  
1               36325  
2                  52  
3                 255


business q5.csv:


author_name  avg_publication_lifetime
0      Bill Bryson                        -3
1  Bryce Courtenay                         0
2      Don DeLillo                         6
3    Douglas Adams                        -4
4  Edward P. Jones                         2


business q6.csv:


authors  total_books  avg_rating  min_rating  \
0                  Frank Herbert            5      3.8700        3.84   
1  Hiromu Arakawa/Akira Watanabe            5      4.5400        4.50   
2             Robert A. Heinlein            5      3.9100        3.78   
3    J.K. Rowling/Mary GrandPré            4      4.6000        4.49   
4                   Henry Miller            4      3.9225        3.78   

   max_rating  std_dev_rating  avg_rating_increase  avg_rating_decrease  \
0        3.91        0.023664             0.000000            -0.012500   
1        4.57        0.026077             0.013333             0.015000   
2        4.01        0.080250             0.013333             0.035000   
3        4.78        0.108397             0.040000             0.003333   
4        4.10        0.126170             0.025000            -0.050000   

   earliest_book_rating  latest_book_rating  
0                  3.91                3.86  
1                  4.50                4.56  
2                  3.87                4.01  
3                  4.56                4.57  
4                  3.98                3.83


business q7.csv:


num_pages  page_bin  bin_avg_ratings  bin_avg_text_reviews  \
0        103         0         3.887346             1012.9938   
1        197         0         3.887346             1012.9938   
2        311         0         3.887346             1012.9938   
3        208         0         3.887346             1012.9938   
4        192         0         3.887346             1012.9938   

   total_books_by_author  avg_author_rating  rating_rank_in_bin  \
0                      1               3.87                   1   
1                      1               3.86                   2   
2                      2               3.83                   3   
3                      2               3.80                   4   
4                      5               4.54                   5   

   lang_avg_ratings  lang_avg_reviews  
0     102504.076344       2196.958668  
1     102504.076344       2196.958668  
2     102504.076344       2196.958668  
3     102504.076344       2196.958668  
4     102504.076344       2196.958668


business q8 final.csv:


title  \
0            The Power of One (The Power of One  #1)   
1               Poetry for Young People: Edward Lear   
2                                  Gravity's Rainbow   
3  Growing Up Untouchable in India: A Dalit Autob...   
4  The Player's Handbook: The Ultimate Guide on D...   

                                            authors  average_rating  \
0                                   Bryce Courtenay            4.35   
1  Edward Lear/Laura Huliska-Beith/Edward Mendelson            3.91   
2                                    Thomas Pynchon            4.01   
3           Vasant Moon/Gail Omvedt/Eleanor Zelliot            3.65   
4                       Heidi Fleiss/Libby Keatinge            3.82   

   ratings_count  text_reviews_count  review_rating_ratio  polarity_rank  
0             45                  13              0.28889              1  
1             66                  19              0.28788              2  
2            762                 213              0.27953              3  
3             16                   4              0.25000              4  
4             34                   8              0.23529              5


business q8 maybe.csv:


isbn  avg_rating  num_editions
0  439785960        4.57             1
1  439358078        4.49             1
2  439554896        4.42             1
3   43965548        4.56             1
4  439682584        4.78             1


Cancer.csv:


id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]


Churn_Modelling.csv:


RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0


CO2 emissions per capita per country.csv:


Country Name Country Code      1960      1961      1962      1963      1964  \
0        Aruba          ABW       NaN       NaN       NaN       NaN       NaN   
1  Afghanistan          AFG  0.046060  0.053604  0.073765  0.074233  0.086292   
2       Angola          AGO  0.097472  0.079038  0.201289  0.192535  0.201003   
3      Albania          ALB  1.258195  1.374186  1.439956  1.181681  1.111742   
4      Andorra          AND       NaN       NaN       NaN       NaN       NaN   

       1965      1966      1967  ...       2009       2010       2011  \
0       NaN       NaN       NaN  ...  25.915833  24.670529  24.505835   
1  0.101467  0.107637  0.123734  ...   0.241723   0.293837   0.412017   
2  0.191528  0.246413  0.154912  ...   1.232495   1.243406   1.252789   
3  1.166099  1.333055  1.363746  ...   1.495600   1.578574   1.803715   
4       NaN       NaN       NaN  ...   6.121652   6.122595   5.867130   

        2012      2013      2014  2015  2016  2017  2018  
0  13.155542  8.351294  8.408363   NaN   NaN   NaN   NaN  
1   0.350371  0.315602  0.299445   NaN   NaN   NaN   NaN  
2   1.330843  1.254617  1.291328   NaN   NaN   NaN   NaN  
3   1.692908  1.749211  1.978763   NaN   NaN   NaN   NaN  
4   5.916597  5.900753  5.832170   NaN   NaN   NaN   NaN  

[5 rows x 61 columns]


CO2emissions.csv:


Country Name Country Code      1960      1961      1962      1963      1964  \
0        Aruba          ABW       NaN       NaN       NaN       NaN       NaN   
1  Afghanistan          AFG  0.046060  0.053604  0.073765  0.074233  0.086292   
2       Angola          AGO  0.097472  0.079038  0.201289  0.192535  0.201003   
3      Albania          ALB  1.258195  1.374186  1.439956  1.181681  1.111742   
4      Andorra          AND       NaN       NaN       NaN       NaN       NaN   

       1965      1966      1967  ...       2009       2010       2011  \
0       NaN       NaN       NaN  ...  25.915833  24.670529  24.505835   
1  0.101467  0.107637  0.123734  ...   0.241723   0.293837   0.412017   
2  0.191528  0.246413  0.154912  ...   1.232495   1.243406   1.252789   
3  1.166099  1.333055  1.363746  ...   1.495600   1.578574   1.803715   
4       NaN       NaN       NaN  ...   6.121652   6.122595   5.867130   

        2012      2013      2014  2015  2016  2017  2018  
0  13.155542  8.351294  8.408363   NaN   NaN   NaN   NaN  
1   0.350371  0.315602  0.299445   NaN   NaN   NaN   NaN  
2   1.330843  1.254617  1.291328   NaN   NaN   NaN   NaN  
3   1.692908  1.749211  1.978763   NaN   NaN   NaN   NaN  
4   5.916597  5.900753  5.832170   NaN   NaN   NaN   NaN  

[5 rows x 61 columns]


CorporateFinancials.csv:


Year  Quarter  Dividend  After_Tax_Profit
0  1997        1     320.5             599.1
1  1997        2     330.6             619.4
2  1997        3     339.7             641.5
3  1997        4     347.1             629.6
4  1998        1     350.1             562.9


csvtest.csv:


Index  C1  C2  C3  C4  C5  C6  C7
0      1   2   3   2   3  29   7  21
1      2   2   5   3   3  23   7  21
2      3   1   3   3   4  23   7  27
3      4   2   3   2   3  34   5  13


data_week.csv:


yearweek  id_week  netsales_online  direct_mail   emails  phone_calls  \
0    201501        1         36999601      1657409  7940890        60289   
1    201502        2         48835239      1636261  8006703       124623   
2    201503        3         36129866       110936  5459639        99061   
3    201504        4         38837398      4438913  8244171        88189   
4    201505        5         37626122        64813  5524945        75075   

   online_budget  promotion  holiday  avg_cloud_index  avg_temp  \
0     1592882.70          1        1         6.389505  1.511591   
1     1532079.97          1        0         6.771789  5.759182   
2     1422730.83          0        0         6.550906  1.422229   
3     1383643.41          0        0         7.137731  0.396899   
4     1388096.99          0        0         6.561078 -0.710233   

   avg_wind_speed  avg_rainfall_level  avg_hours_sun  avg_snow_level  
0        4.818268            2.139169       1.694728        4.352802  
1        7.652616            4.892557       1.054485        1.243305  
2        3.280275            1.417072       1.263986        1.249276  
3        3.905749            2.244240       0.582608        2.949619  
4        3.804411            1.946786       1.648585        8.217035


departments new.csv:


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol


departments.csv:


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol


Event_entry.csv:


Entry  TicketClass     Sex   Age  Price  Order  \
0      0            3    male  22.0      7      1   
1      1            1  female  38.0     71      2   
2      1            3  female  26.0      8      3   
3      1            1  female  35.0     53      4   
4      0            3    male  35.0      8      5   

                                                Name  Ticket Number  \
0                            Braund, Mr. Owen Harris         648155   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...         119239   
2                             Heikkinen, Miss. Laina         411408   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)         530093   
4                           Allen, Mr. William Henry         510583   

   Bank Balance  
0         74389  
1         94932  
2         26146  
3         35627  
4         60457


fundamentals.csv:


Unnamed: 0 Ticker Symbol Period Ending  Accounts Payable  \
0           0           AAL    2012-12-31      3.068000e+09   
1           1           AAL    2013-12-31      4.975000e+09   
2           2           AAL    2014-12-31      4.668000e+09   
3           3           AAL    2015-12-31      5.102000e+09   
4           4           AAP    2012-12-29      2.409453e+09   

   Accounts Receivable  Add'l income/expense items  After Tax ROE  \
0         -222000000.0               -1.961000e+09           23.0   
1          -93000000.0               -2.723000e+09           67.0   
2         -160000000.0               -1.500000e+08          143.0   
3          352000000.0               -7.080000e+08          135.0   
4          -89482000.0                6.000000e+05           32.0   

   Capital Expenditures  Capital Surplus  Cash Ratio  ...  \
0         -1.888000e+09     4.695000e+09        53.0  ...   
1         -3.114000e+09     1.059200e+10        75.0  ...   
2         -5.311000e+09     1.513500e+10        60.0  ...   
3         -6.151000e+09     1.159100e+10        51.0  ...   
4         -2.711820e+08     5.202150e+08        23.0  ...   

   Total Current Assets  Total Current Liabilities  Total Equity  \
0          7.072000e+09               9.011000e+09 -7.987000e+09   
1          1.432300e+10               1.380600e+10 -2.731000e+09   
2          1.175000e+10               1.340400e+10  2.021000e+09   
3          9.985000e+09               1.360500e+10  5.635000e+09   
4          3.184200e+09               2.559638e+09  1.210694e+09   

   Total Liabilities  Total Liabilities & Equity  Total Revenue  \
0       2.489100e+10                1.690400e+10   2.485500e+10   
1       4.500900e+10                4.227800e+10   2.674300e+10   
2       4.120400e+10                4.322500e+10   4.265000e+10   
3       4.278000e+10                4.841500e+10   4.099000e+10   
4       3.403120e+09                4.613814e+09   6.205003e+09   

   Treasury Stock  For Year  Earnings Per Share  Estimated Shares Outstanding  
0    -367000000.0    2012.0               -5.60                  3.350000e+08  
1             0.0    2013.0              -11.25                  1.630222e+08  
2             0.0    2014.0                4.02                  7.169154e+08  
3             0.0    2015.0               11.39                  6.681299e+08  
4     -27095000.0    2012.0                5.29                  7.328355e+07  

[5 rows x 79 columns]


Gillette_data.csv:


year     Revenues  Sqrt_Advtg  Lag.Revenues    Margin  Sales_Multiple
0  1975  4939.861401   25.541576   4775.844134  0.188893        0.713518
1  1976  4951.590218   25.144198   4939.861401  0.179681        0.562394
2  1977  4947.587130   25.457227   4951.590218  0.181214        0.464744
3  1978  4955.640844   26.363812   4947.587130  0.195540        0.455789
4  1979  5164.101886   27.108974   4955.640844  0.198043        0.400943


GlobalLandTemperatures_GlobalLandTemperaturesByCountry.csv:


dt  AverageTemperature  AverageTemperatureUncertainty Country
0  1743-11-01               4.384                          2.294   Åland
1  1743-12-01                 NaN                            NaN   Åland
2  1744-01-01                 NaN                            NaN   Åland
3  1744-02-01                 NaN                            NaN   Åland
4  1744-03-01                 NaN                            NaN   Åland


hockey_data_full.csv:


S6v5  S6v4  S6v3  S5v4  S5v3  S4v3  SNG  ANA.20022003  ATL.20022003  \
0     0     0     0     0     0     0    0             0             0   
1     0     0     0     0     0     0    0             0             0   
2     0     0     0     0     0     0    0             0             0   
3     0     0     0    -1     0     0    0             0             0   
4     0     0     0     0     1     0    0             0             0   

   BOS.20022003  ...  KENNY_AGOSTINO  ANDREW_CAMPBELL  ADAM_PAYERL  \
0             0  ...               0                0            0   
1             0  ...               0                0            0   
2             0  ...               0                0            0   
3             0  ...               0                0            0   
4             0  ...               0                0            0   

   SCOTT_MAYFIELD  CEDRIC_PAQUETTE  JUSTIN_JOHNSON  TY_RATTIE  CONNOR_KNAPP  \
0               0                0               0          0             0   
1               0                0               0          0             0   
2               0                0               0          0             0   
3               0                0               0          0             0   
4               0                0               0          0             0   

   MATT_CAREY  JOEY_HISHON  
0           0            0  
1           0            0  
2           0            0  
3           0            0  
4           0            0  

[5 rows x 2776 columns]


hockey_data_y.csv:


y
0  0
1  0
2  1
3  0
4  1


Melbourne_housing.csv:


Suburb         Address  Rooms Type      Price Method   SellerG  \
0          Thornbury  7/67 Pender St      2    u   438000.0     SP      Love   
1             Coburg      32 Rose St      2    h   901000.0      S     Peter   
2     Bentleigh East  2/25 Brooks St      3    u   800000.0      S  Woodards   
3  Templestowe Lower    2/35 John St      3    t  1155000.0      S    Jellis   
4        South Yarra   1/35 Marne St      3    u  1630000.0     PI    Jellis   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  22/08/2016       6.5      3071  ...       1.0  1.0       0.0          66.0   
1  27/06/2016       7.8      3058  ...       1.0  1.0     545.0         107.0   
2  28/05/2016      13.9      3165  ...       2.0  2.0     261.0           NaN   
3  25/11/2017      12.4      3107  ...       NaN  NaN       NaN           NaN   
4   10/9/2016       3.3      3141  ...       1.0  1.0       0.0           NaN   

   YearBuilt              CouncilArea Lattitude  Longtitude  \
0     1970.0     Darebin City Council  -37.7541    145.0088   
1     1940.0     Darebin City Council  -37.7445    144.9471   
2        NaN   Glen Eira City Council  -37.9163    145.0779   
3        NaN  Manningham City Council       NaN         NaN   
4        NaN   Melbourne City Council  -37.8359    144.9839   

              Regionname Propertycount  
0  Northern Metropolitan          8870  
1  Northern Metropolitan         11204  
2  Southern Metropolitan         10969  
3   Eastern Metropolitan          5420  
4  Southern Metropolitan         14887  

[5 rows x 21 columns]


mock_data.csv:


First Name Last Name  Gender  Age                     Email        Phone  \
0      Aston  Johnston    Male   18      a.johnston@gmail.com  542-6763-58   
1      Sarah   Stevens  Female   18  s.stevens@randatmail.com  263-5909-20   
2    Dominik     Clark    Male   18    d.clark@randatmail.com  514-8169-50   
3      Lenny   Chapman    Male   30  l.chapman@randatmail.com  453-0589-04   
4    William   Stevens    Male   26  w.stevens@randatmail.com  395-2566-58   

  Education        Occupation  Experience (Years)  Salary Marital Status  \
0  Doctoral       Interpreter                   4  164702         Single   
1   Primary      Veterinarian                   9  184108         Single   
2   Primary  Fashion Designer                   6  122914         Single   
3    Master             Medic                   6  189918         Single   
4  Doctoral        Astronomer                   8   96976         Single   

               IP   Last Seen  Number of Children        SSN    Password  \
0    46.60.226.75  08/23/2012                   2  122725502  kovyheb0jl   
1  201.212.238.96  01/16/2001                   2  711208210  g8937x5d0h   
2    97.129.17.52  05/04/2022                   3   58987219  hvn6je561h   
3  199.199.98.208  12/04/2012                   3  535367549  7af8nqe6ns   
4    134.100.31.5  12/05/2016                   4  708473341  o50nw2bmbf   

   Number of Logins  
0                38  
1                53  
2                80  
3                72  
4                13


orders new.csv:


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0


orders.csv:


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0


order_products__prior new.csv:


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0


order_products__prior.csv:


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0


order_products__train new.csv:


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1


order_products__train.csv:


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1


pickups.csv:


year   miles  price   make
0  2008   17638  14995    GMC
1  2003  174000   8500  Dodge
2  2001    1500   9998  Dodge
3  2007   22422  23950    GMC
4  2007   34815  19980    GMC


products new.csv:


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13


products.csv:


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13


publisher.csv:


publisher  bookID
0  Scholastic Inc.       1
1  Scholastic Inc.       2
2       Scholastic       4
3  Scholastic Inc.       5
4       Scholastic       8


Qualcomm_CBC_Design.csv:


Version (0 = Fixed Task)  Task  Concept  Att 1 - Brand  \
0                         0     1        1            NaN   
1                         0     1        2            NaN   
2                         0     1        3            NaN   
3                         0     1        4            NaN   
4                         0     2        1            NaN   

   Att 2 - Operating system  Att 3 - CPU  Att 4 - Price  Att 5 - AI Capability  
0                       NaN          NaN            NaN                    NaN  
1                       NaN          NaN            NaN                    NaN  
2                       NaN          NaN            NaN                    NaN  
3                       NaN          NaN            NaN                    NaN  
4                       NaN          NaN            NaN                    NaN


ratings.csv:


average_rating  ratings_count  bookID
0           4.57        2095690       1
1           4.49        2153167       2
2           4.42           6333       4
3           4.56        2339585       5
4           4.78          41428       8


sample_submission new.csv:


order_id     products
0        17  39276 29259
1        34  39276 29259
2       137  39276 29259
3       182  39276 29259
4       257  39276 29259


sample_submission.csv:


order_id     products
0        17  39276 29259
1        34  39276 29259
2       137  39276 29259
3       182  39276 29259
4       257  39276 29259


sommer_deger.csv:


instrument  treatment  outcome
0           0          0        0
1           0          0        0
2           0          0        0
3           0          0        0
4           0          0        0


Telco-Customer-Churn.csv:


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]


text_reviews.csv:


text_reviews_count  bookID
0               27591       1
1               29221       2
2                 244       4
3               36325       5
4                 164       8


Titanic.csv:


Name PClass    Age     Sex  \
0                   Allen, Miss Elisabeth Walton    1st  29.00  female   
1                    Allison, Miss Helen Loraine    1st   2.00  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                  Allison, Master Hudson Trevor    1st   0.92    male   

   Survived  
0         1  
1         0  
2         0  
3         0  
4         1


Toggl_time_entries_2024-01-15_to_2024-01-21 (1).csv:


User                  Email    Client          Project  Task  \
0  Shalagha  spmundepi@ucdavis.edu  Qualcomm  Survey Analysis   NaN   

                                         Description Billable  Start date  \
0  Create a report on market research questions t...       No  2024-01-20   

  Start time    End date  End time  Duration  Tags  Amount (USD)  
0   13:28:19  2024-01-20  13:28:34  00:00:15   NaN           NaN


Toggl_time_entries_2024-02-05_to_2024-02-11 (1).csv:


User                  Email  Client  Project  Task  \
0  Shalagha  spmundepi@ucdavis.edu     NaN      NaN   NaN   
1  Shalagha  spmundepi@ucdavis.edu     NaN      NaN   NaN   

                                         Description Billable  Start date  \
0                                     Assignment 3.5       No  2024-02-06   
1  Internal meeting to finalize survey questions ...       No  2024-02-07   

  Start time    End date  End time  Duration  Tags  Amount (USD)  
0   02:41:38  2024-02-06  08:02:26  05:20:48   NaN           NaN  
1   19:22:34  2024-02-07  19:56:11  00:33:37   NaN           NaN


Traffic_Crashes_Resulting_in_Injury_20240407.csv:


C:\Users\shala\AppData\Local\Temp\ipykernel_9072\1599710358.py:8: DtypeWarning: Columns (3,15) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file)


unique_id  cnn_intrsctn_fkey  cnn_sgmt_fkey case_id_pkey  tb_latitude  \
0      73551         22882000.0      4060000.0    230884432    37.743586   
1      23055         23838000.0            NaN    150060071    37.781750   
2      41755         27803000.0      1977000.0    130332135    37.761830   
3      24743         30739000.0      8747203.0      3348111    37.783992   
4      30142         26494000.0      6039101.0      4395868    37.785291   

   tb_longitude   geocode_source geocode_location      collision_datetime  \
0   -122.463710  SFPD-INTERIM DB      CITY STREET  12/18/2023 02:15:00 PM   
1   -122.396380  SFPD-CROSSROADS      CITY STREET  01/20/2015 10:07:00 AM   
2   -122.503878  SFPD-CROSSROADS      CITY STREET  04/23/2013 11:26:00 AM   
3   -122.408066  SFPD-CROSSROADS      CITY STREET  08/22/2007 10:30:00 PM   
4   -122.424770  SFPD-CROSSROADS      CITY STREET  09/03/2009 01:35:00 PM   

      collision_date  ... party2_move_pre_acc  \
0   2023 December 18  ...             Backing   
1    2015 January 20  ...   Making Right Turn   
2      2013 April 23  ...              Parked   
3     2007 August 22  ...                 NaN   
4  2009 September 03  ...              Parked   

                                 point              data_as_of  \
0   POINT (-122.46370951 37.743585526)  02/16/2024 12:00:00 AM   
1  POINT (-122.396379519 37.781750283)  01/20/2015 12:00:00 AM   
2  POINT (-122.503877741 37.761830418)  04/23/2013 12:00:00 AM   
3  POINT (-122.408066364 37.783992037)  08/22/2007 12:00:00 AM   
4  POINT (-122.424770266 37.785291007)  09/03/2009 12:00:00 AM   

          data_updated_at          data_loaded_at Neighborhoods  \
0  02/16/2024 12:00:00 AM  02/16/2024 04:12:43 PM          49.0   
1  04/26/2023 12:00:00 AM  02/16/2024 04:12:43 PM          32.0   
2  04/26/2023 12:00:00 AM  02/16/2024 04:12:43 PM          39.0   
3  04/26/2023 12:00:00 AM  02/16/2024 04:12:43 PM          19.0   
4  04/26/2023 12:00:00 AM  02/16/2024 04:12:43 PM         100.0   

  SF Find Neighborhoods Current Police Districts Current Supervisor Districts  \
0                  49.0                     10.0                          8.0   
1                  32.0                      1.0                         10.0   
2                  39.0                     10.0                          7.0   
3                  19.0                      5.0                         10.0   
4                 100.0                      4.0                         11.0   

  Analysis Neighborhoods  
0                   41.0  
1                    8.0  
2                   35.0  
3                   34.0  
4                   39.0  

[5 rows x 60 columns]


updated_conjoint_survey_mock_data.csv:


Current PC Brand Ownership Key Features in Purchasing Decision  \
0                         HP                 Graphics card;Price   
1                      Apple           Portability;Graphics card   
2                      Apple                   Brand;Screen size   
3                         HP           Graphics card;Portability   
4                     Lenovo                     RAM;Portability   

  Awareness of SD Processors Self-Identification Category  \
0                        Yes                   Freelancer   
1                        Yes               Business owner   
2                        Yes         Working professional   
3                         No                      Student   
4                        Yes                      Student   

  Influence of Brand on Purchase Decision Background in Education or Work  \
0                            A great deal                           Other   
1                                Somewhat                Associate degree   
2                            A great deal                Associate degree   
3                            A great deal                     Engineering   
4                                Somewhat               Bachelor's degree   

  Age Range                   Ethnicity State of Residence  Q1_Option_Apple  \
0  Under 18                       Other                 TX                0   
1     35-44   Black or African American                 IL                1   
2  Under 18   Black or African American                 MA                0   
3     35-44                       Asian                 TX                0   
4     35-44                       Other                 IN                0   

   ...  Q8_Option_Intel Core  Q8_Option_AMD Ryzen  Q9_Option_Apple  \
0  ...                     0                    0                0   
1  ...                     0                    0                1   
2  ...                     0                    0                1   
3  ...                     0                    0                0   
4  ...                     0                    1                0   

   Q9_Option_Snapdragon  Q9_Option_Intel Core  Q9_Option_AMD Ryzen  \
0                     1                     0                    0   
1                     0                     0                    0   
2                     0                     0                    0   
3                     1                     0                    0   
4                     1                     0                    0   

   Q10_Option_Apple  Q10_Option_Snapdragon  Q10_Option_Intel Core  \
0                 0                      0                      1   
1                 1                      0                      0   
2                 1                      0                      0   
3                 0                      1                      0   
4                 0                      0                      1   

   Q10_Option_AMD Ryzen  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 49 columns]


x_vars_housing.csv:


Unnamed: 0         0         1         2         3         4         5  \
0           0 -0.034052  0.204160  0.292150  0.050746  0.747918 -0.214101   
1           1  0.684733  0.866117  1.156674  0.050746 -0.673674 -0.214101   
2           2 -0.034052 -0.899102 -0.522785  0.050746  0.747918 -0.214101   
3           3  0.205543 -0.170949  0.353341  0.050746 -1.384470  3.410101   
4           4  1.643112 -1.428667  0.409385  0.050746 -0.673674 -0.214101   

          6         7         8  ...        64        65        66        67  \
0 -0.801718 -0.513137 -0.748987  ...  0.961834 -0.214101  0.747918  0.276075   
1  2.808176  3.086528 -2.716241  ...  0.961834 -0.214101 -0.673674  0.276075   
2 -0.801718  0.386779 -1.630860  ...  0.961834 -0.214101  0.747918  0.276075   
3  0.642240 -0.513137  0.539903  ... -1.543940  3.410101 -1.384470  0.276075   
4  1.364219 -0.513137  0.811248  ... -1.543940 -0.214101 -0.673674  0.276075   

         68        69        70        71        72        73  
0  0.050746  0.252492 -0.034052 -0.513137 -0.619491  0.504414  
1  0.050746  0.252492  0.684733  3.086528 -1.372421  0.504414  
2  0.050746  0.252492 -0.034052  0.386779  1.639299  0.167175  
3  0.050746  0.252492  0.205543 -0.513137 -0.619491 -0.170063  
4  0.050746  0.252492  1.643112 -0.513137  0.133439 -1.181777  

[5 rows x 75 columns]


y_var_housing.csv:


Unnamed: 0  SalePrice
0         296  11.931642
1         185  13.071072
2         380  11.751950
3         208  12.531776
4         400  12.411056

In [4]:
# Read order_products__prior.csv and products.csv
order_products_prior = pd.read_csv(os.path.join(input_path, 'order_products__prior.csv'))
products = pd.read_csv(os.path.join(input_path, 'products.csv'))

In [5]:
print('order_products_prior -- dimensions: {0};   size: {1}'.format(order_products_prior.shape, order_products_prior.size))


order_products_prior -- dimensions: (32434489, 4);   size: 129737956


In [6]:
order_products_prior = order_products_prior.set_index('order_id')['product_id'].rename('item_id')
display(order_products_prior.head(10))
type(order_products_prior)

order_id
2    33120
2    28985
2     9327
2    45918
2    30035
2    17794
2    40141
2     1819
2    43668
3    33754
Name: item_id, dtype: int64

pandas.core.series.Series

In [7]:
print('dimensions: {0};   size: {1};   unique_orders: {2};   unique_items: {3}'
      .format(order_products_prior.shape, order_products_prior.size, len(order_products_prior.index.unique()), len(order_products_prior.value_counts())))


dimensions: (32434489,);   size: 32434489;   unique_orders: 3214874;   unique_items: 49677


In [13]:
class AprioriAlg():
    # Returns frequency counts for items and item pairs
    def freq(self, iterable):
        if type(iterable) == pd.core.series.Series:
            return iterable.value_counts().rename("freq")
        else: 
            return pd.Series(Counter(iterable)).rename("freq")


    # Returns number of unique orders
    def order_count(self,order_item):
        return len(set(order_item.index))


    # Returns generator that yields item pairs, one at a time
    def get_item_pairs(self,order_item):
        order_item = order_item.reset_index().to_numpy()
        for order_id, order_object in groupby(order_item, lambda x: x[0]):
            item_list = [item[1] for item in order_object]

            for item_pair in combinations(item_list, 2):
                yield item_pair


    # Returns frequency and support associated with item
    def merge_item_stats(self,item_pairs, item_stats):
        return (item_pairs
                    .merge(item_stats.rename(columns={'freq': 'freqA', 'support': 'supportA'}), left_on='item_A', right_index=True)
                    .merge(item_stats.rename(columns={'freq': 'freqB', 'support': 'supportB'}), left_on='item_B', right_index=True))


    # Returns name associated with item
    def merge_item_name(self,rules, item_name):
        columns = ['itemA','itemB','freqAB','supportAB','freqA','supportA','freqB','supportB', 
                   'confidenceAtoB','confidenceBtoA','lift']
        rules = (rules
                    .merge(item_name.rename(columns={'item_name': 'itemA'}), left_on='item_A', right_on='item_id')
                    .merge(item_name.rename(columns={'item_name': 'itemB'}), left_on='item_B', right_on='item_id'))
        return rules[columns] 
    
    # Return association rules sorted by lift in descending order
    def fit_transform(self,order_item, min_support):
        print("Commencing item count: {:22d}".format(len(order_item)))

        # Calculate item frequency and support
        item_stats = self.freq(order_item).to_frame("freq")
        item_stats['support'] = item_stats['freq'] / self.order_count(order_item) * 100

        # Filter items below min support 
        qualifying_items = item_stats[item_stats['support'] >= min_support].index
        order_item = order_item[order_item.isin(qualifying_items)]

        print("Items with support >= {}: {:15d}".format(min_support, len(qualifying_items)))
        print("Remaining order_item: {:21d}".format(len(order_item)))

        # Filter orders with less than 2 items
        order_size = self.freq(order_item.index)
        qualifying_orders = order_size[order_size >= 2].index
        order_item = order_item[order_item.index.isin(qualifying_orders)]

        print("Remaining orders with 2+ items: {:11d}".format(len(qualifying_orders)))
        print("Remaining order_item: {:21d}".format(len(order_item)))

        # Recalculate item frequency and support
        item_stats = self.freq(order_item).to_frame("freq")
        item_stats['support'] = item_stats['freq'] / self.order_count(order_item) * 100

        # Get item pairs generator
        item_pair_gen = self.get_item_pairs(order_item)

        # Calculate item pair frequency and support
        item_pairs = self.freq(item_pair_gen).to_frame("freqAB")
        item_pairs['supportAB'] = item_pairs['freqAB'] / len(qualifying_orders) * 100

        print("Identified item pairs: {:31d}".format(len(item_pairs)))

        # Filter item pairs below min support
        item_pairs = item_pairs[item_pairs['supportAB'] >= min_support]

        print("Item pairs with support >= {}: {:10d}\n".format(min_support, len(item_pairs)))

        # Create table of association rules and compute relevant metrics
        item_pairs = item_pairs.reset_index().rename(columns={'level_0': 'item_A', 'level_1': 'item_B'})
        item_pairs = self.merge_item_stats(item_pairs, item_stats)

        item_pairs['confidenceAtoB'] = item_pairs['supportAB'] / item_pairs['supportA']
        item_pairs['confidenceBtoA'] = item_pairs['supportAB'] / item_pairs['supportB']
        item_pairs['lift'] = item_pairs['supportAB'] / (item_pairs['supportA'] * item_pairs['supportB'])

        return item_pairs.sort_values('lift', ascending=False)


In [9]:
# Instantiate the AprioriAlg class
alg = AprioriAlg()

# Use only a portion of the data (1/3)
items = order_products_prior[:len(order_products_prior) // 3]

# Apply the Apriori algorithm to find association rules with minimum support of 0.01
rules = alg.fit_transform(items, min_support=0.01)

Starting order_item:               10811496
Items with support >= 0.01:           10915
Remaining order_item:               9953094
Remaining orders with 2+ items:     1004690
Remaining order_item:               9892535
Item pairs:                        17751273
Item pairs with support >= 0.01:      48935



In [10]:
# Rename product_id to item_id and product_name to item_name
item_names = products.rename(columns={'product_id': 'item_id', 'product_name': 'item_name'})

# Merge item names with association rules and sort by lift in descending order
rules_final = alg.merge_item_name(rules, item_names).sort_values('lift', ascending=False)

# Reset index and display the association rules
rules_final = rules_final.reset_index(drop=True)
display(rules_final)

itemA  \
0       Organic Strawberry Chia Lowfat 2% Cottage Cheese   
1                    Grain Free Chicken Formula Cat Food   
2         Organic Grapefruit Ginger Sparkling Yerba Mate   
3              Organic Fruit Yogurt Smoothie Mixed Berry   
4      Nonfat Strawberry With Fruit On The Bottom Gre...   
...                                                  ...   
48930                               Organic Strawberries   
48931                               Organic Hass Avocado   
48932                                    Organic Avocado   
48933                                             Banana   
48934                             Bag of Organic Bananas   

                                            itemB  freqAB  supportAB   freqA  \
0      Organic Cottage Cheese Blueberry Acai Chia     118   0.011745     412   
1              Grain Free Turkey Formula Cat Food     107   0.010650     641   
2      Cranberry Pomegranate Sparkling Yerba Mate     125   0.012442     557   
3           Apple Blueberry Fruit Yogurt Smoothie     121   0.012044     494   
4        0% Greek, Blueberry on the Bottom Yogurt     126   0.012541     541   
...                                           ...     ...        ...     ...   
48930                                Strawberries     210   0.020902   88235   
48931                             Organic Avocado     147   0.014631   70883   
48932                        Organic Hass Avocado     137   0.013636   58400   
48933                      Bag of Organic Bananas     214   0.021300  157187   
48934                                      Banana     173   0.017219  125715   

        supportA   freqB   supportB  confidenceAtoB  confidenceBtoA      lift  
0       0.041008     300   0.029860        0.286408        0.393333  9.591701  
1       0.063801     286   0.028466        0.166927        0.374126  5.863971  
2       0.055440     389   0.038718        0.224417        0.321337  5.796119  
3       0.049169     441   0.043894        0.244939        0.274376  5.580228  
4       0.053847     445   0.044292        0.232902        0.283146  5.258300  
...          ...     ...        ...             ...             ...       ...  
48930   8.782311   47133   4.691298        0.002380        0.004455  0.000507  
48931   7.055211   58400   5.812738        0.002074        0.002517  0.000357  
48932   5.812738   70883   7.055211        0.002346        0.001933  0.000333  
48933  15.645323  125715  12.512815        0.001361        0.001702  0.000109  
48934  12.512815  157187  15.645323        0.001376        0.001101  0.000088  

[48935 rows x 11 columns]

In [11]:
rules_final.loc[rules_final['itemA'] == 'Organic Hass Avocado'] 

itemA                             itemB  freqAB  \
3314   Organic Hass Avocado     Organic Large Green Asparagus    1121   
3590   Organic Hass Avocado         Organic Red Cherry Tomato     163   
3612   Organic Hass Avocado          Tortillas, Corn, Organic     341   
3806   Organic Hass Avocado  Wild Albacore Tuna No Salt Added     188   
3833   Organic Hass Avocado   Organic Large White Flesh Peach     265   
...                     ...                               ...     ...   
48759  Organic Hass Avocado                           Spinach     108   
48852  Organic Hass Avocado                       Gala Apples     134   
48859  Organic Hass Avocado       Organic 2% Reduced Fat Milk     128   
48893  Organic Hass Avocado                        Whole Milk     167   
48931  Organic Hass Avocado                   Organic Avocado     147   

       supportAB  freqA  supportA  freqB  supportB  confidenceAtoB  \
3314    0.111577  70883  7.055211   6398  0.636813        0.015815   
3590    0.016224  70883  7.055211    985  0.098040        0.002300   
3612    0.033941  70883  7.055211   2068  0.205835        0.004811   
3806    0.018712  70883  7.055211   1178  0.117250        0.002652   
3833    0.026376  70883  7.055211   1670  0.166220        0.003739   
...          ...    ...       ...    ...       ...             ...   
48759   0.010750  70883  7.055211   5652  0.562562        0.001524   
48852   0.013337  70883  7.055211   7966  0.792881        0.001890   
48859   0.012740  70883  7.055211   7771  0.773472        0.001806   
48893   0.016622  70883  7.055211  11780  1.172501        0.002356   
48931   0.014631  70883  7.055211  58400  5.812738        0.002074   

       confidenceBtoA      lift  
3314         0.175211  0.024834  
3590         0.165482  0.023455  
3612         0.164894  0.023372  
3806         0.159593  0.022621  
3833         0.158683  0.022492  
...               ...       ...  
48759        0.019108  0.002708  
48852        0.016821  0.002384  
48859        0.016471  0.002335  
48893        0.014177  0.002009  
48931        0.002517  0.000357  

[1147 rows x 11 columns]

In [ ]:
Part A - Question 2

Here are two business recommendations that can help deliver effective business outcomes by leveraging the insights from our association rule-based recommender system:

1. Targeted Promotion and Upselling Strategies
Insight: Certain product pairs show a high degree of association, such as "Organic Strawberry Chia Lowfat 2% Cottage Cheese" and "Organic Cottage Cheese Blueberry Acai Chia" with a lift of approximately 9.59. This indicates that customers who purchase one are highly likely to purchase the other.

Recommendation:

a. Cross-Selling on Product Pages: Implement a feature on the online platform that suggests complementary products when a customer views or adds an item to their cart. For example, if a customer adds "Organic Strawberry Chia Lowfat 2% Cottage Cheese" to their shopping cart, the system should automatically suggest "Organic Cottage Cheese Blueberry Acai Chia."

b. Targeted Email Marketing: Send personalized email recommendations to customers based on their purchase history, suggesting highly associated products they haven't bought yet but are likely to be interested in.

2. Dynamic Bundling Offers
Insight: Several product pairs like "Grain Free Chicken Formula Cat Food" and "Grain Free Turkey Formula Cat Food" show significant association but may not be bought frequently because of high individual prices or lack of awareness about their complementary nature.

Recommendation:

a. Product Bundling: Offer these items as a bundle at a discounted rate. Bundling products that are frequently purchased together enhances customer convenience and increases the transaction size. For instance, creating a bundle offer for cat food varieties that are frequently purchased together could attract pet owners looking for variety and value.

b. Incentivize Bulk Purchases: Provide a discount or loyalty points when customers purchase both items in a set period. This strategy not only boosts sales but also encourages repeat visits and customer loyalty.